File: mergeSort.c
#include <stdio.h>
#include <stdlib.h>

void printArray(int *a,int n);
void merge(int *A,int *L,int *R,int sizeOfLeft,int sizeOfRight);
void sort(int *arr,int n);

int main(){
    int arr[]={7,3,4,5,2,9,6,8};
    int size=sizeof(arr)/sizeof(arr[0]);
    sort(arr,size);
    printArray(arr,size);
}

void merge(int *A,int *L,int *R,int sizeOfLeft,int sizeOfRight){
    int i=0,j=0,k=0;

    while(i<sizeOfLeft && j<sizeOfRight){

        if(L[i]<R[j]) A[k++]=L[i++];

        else A[k++]=R[j++];
    }

    while(i<sizeOfLeft)A[k++]=L[i++];
    while(j<sizeOfRight) A[k++]=R[j++];
}

void sort(int *arr,int n){
    int mid = n/2;

    if(n<2) return;

    int *L = (int*) malloc(sizeof(int)*mid);
    int *R = (int*) malloc(sizeof(int)*n-mid);

    for(int i=0;i<mid;i++) L[i]=arr[i];
    for(int j=0;j<n-mid;j++) R[j]=arr[j+mid];

    sort(L,mid);
    sort(R,n-mid);

    merge(arr,L,R,mid,n-mid);

    free(L);
    free(R);
}

void printArray(int *a,int n){
    for(int i=0;i<n;i++) printf(" %d ",a[i]);
}

File: MergeSort.java
class MergeSort{
	public static void main(String[] args){
		int[] a={5,4,3,2,1};
		sort(a);
		printArr(a);
	
	}
	public static void sort(int[] a){
		int mid = (a.length)/2;
		if(a.length<2) return;

		int[] L = new int[mid];
		int[] R = new int[a.length-mid];

		for(int i=0;i<mid;i++) L[i]=a[i];
		for(int j=0;j<a.length-mid;j++) R[j]=a[j+mid];
		printArr(L);
		printArr(R);
		sort(L);
		sort(R);

		merge(a,L,R);
		printArr(a);
	}
	public static void merge(int[] a,int[] L,int[] R){
		int i=0,j=0,k=0;
		while(i<L.length && j<R.length){
			if(L[i]<R[j]) a[k++]=L[i++];

			else a[k++]=R[j++];
		}
		
		while(i<L.length) a[k++]=L[i++];
		while(j<R.length) a[k++]=R[j++];
	}
	public static void printArr(int[] a){
		for(int i=0;i<a.length;i++){
			System.out.print(a[i]+" ");
		}
		System.out.println(" ");
	}
}